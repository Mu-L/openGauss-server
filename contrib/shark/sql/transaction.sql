create schema transaction_test;
set search_path = 'transaction_test';

create table t1(c1 int);

begin tran;
insert into t1 values(1);
select * from t1 order by c1;
commit tran;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(2);
select * from t1 order by c1;
commit tran transaction1;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(3);
select * from t1 order by c1;
commit tran transaction2;

begin tran transaction1;
insert into t1 values(4);
select * from t1 order by c1;
commit tran;

begin tran;
insert into t1 values(5);
select * from t1 order by c1;
commit tran transaction2;

begin tran;
insert into t1 values(6);
select * from t1 order by c1;
rollback tran;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(7);
select * from t1 order by c1;
rollback tran;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(8);
save tran savepoint1;
insert into t1 values(9);
rollback tran savePoint1;
select * from t1 order by c1;
commit tran;

begin tran transaction1;
insert into t1 values(10);
save tran savepoint1;
insert into t1 values(11);
rollback tran;
select * from t1 order by c1;

begin tran;
save tran savePoint1;
insert into t1 values(12);
select * from t1 order by c1;
rollback tran savePoint1;
select * from t1 order by c1;
save tran savePoint2;
insert into t1 values(13);
save tran savePoint3;
insert into t1 values(14);
save tran savePoint4;
insert into t1 values(15);
select * from t1 order by c1;
rollback tran savePoint4;
select * from t1 order by c1;
rollback tran savePoint3;
select * from t1 order by c1;
rollback tran savePoint2;
select * from t1 order by c1;
commit tran;
select * from t1 order by c1;

begin tran;
save tran savePoint1;
drop table t1;
\d t1;
rollback tran savePoint1;
select * from t1 order by c1;
save tran savePoint5;
insert into t1 values(16);
save tran savePoint6;
insert into t1 values(17);
save tran savePoint7;
insert into t1 values(18);
select * from t1 order by c1;
rollback tran savePoint6;
select * from t1 order by c1;
commit tran;
select * from t1 order by c1;

begin tran;
insert into t1 values(19);
save tran savePoint8;
insert into t1 values(20);
save tran savePoint9;
insert into t1 values(21);
save tran savePoint10;
insert into t1 values(22);
select * from t1 order by c1;
rollback tran savePoint8;
select * from t1 order by c1;
commit tran;
select * from t1 order by c1;

begin tran transaction1;
save tran savePoint1;
insert into t1 values(23);
select * from t1 order by c1;
-- expect error
rollback tran savePoint2;
select * from t1 order by c1;
rollback tran;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(24);
begin tran transaction2;
insert into t1 values(25);
select * from t1 order by c1;
commit tran transaction2;
select * from t1 order by c1;
commit tran transaction1;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(26);
begin tran transaction2;
insert into t1 values(27);
select * from t1 order by c1;
rollback tran;
select * from t1 order by c1;
rollback tran;
select * from t1 order by c1;

begin tran transaction1;
insert into t1 values(28);
begin tran transaction2;
insert into t1 values(29);
select * from t1 order by c1;
commit tran transaction1;
select * from t1 order by c1;
commit tran transaction2;
select * from t1 order by c1;

truncate table t1;
begin tran;
insert into t1 values(1);
commit work;
select * from t1;

begin tran;
insert into t1 values(2);
rollback work;
select * from t1;

begin tran;
insert into t1 values(3);
commit;
select * from t1;

begin tran;
insert into t1 values(4);
rollback;
select * from t1;

truncate table t1;
begin transaction;
insert into t1 values(1);
select * from t1 order by c1;
commit transaction;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(2);
select * from t1 order by c1;
commit transaction transaction1;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(3);
select * from t1 order by c1;
commit transaction transaction2;

begin transaction transaction1;
insert into t1 values(4);
select * from t1 order by c1;
commit transaction;

begin transaction;
insert into t1 values(5);
select * from t1 order by c1;
commit transaction transaction2;

begin transaction;
insert into t1 values(6);
select * from t1 order by c1;
rollback transaction;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(7);
select * from t1 order by c1;
rollback transaction;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(8);
save transaction savepoint1;
insert into t1 values(9);
rollback transaction savePoint1;
select * from t1 order by c1;
commit transaction;

begin transaction transaction1;
insert into t1 values(10);
save transaction savepoint1;
insert into t1 values(11);
rollback transaction;
select * from t1 order by c1;

begin transaction;
save transaction savePoint1;
insert into t1 values(12);
select * from t1 order by c1;
rollback transaction savePoint1;
select * from t1 order by c1;
save transaction savePoint2;
insert into t1 values(13);
save transaction savePoint3;
insert into t1 values(14);
save transaction savePoint4;
insert into t1 values(15);
select * from t1 order by c1;
rollback transaction savePoint4;
select * from t1 order by c1;
rollback transaction savePoint3;
select * from t1 order by c1;
rollback transaction savePoint2;
select * from t1 order by c1;
commit transaction;
select * from t1 order by c1;

begin transaction;
save transaction savePoint1;
drop table t1;
\d t1;
rollback transaction savePoint1;
select * from t1 order by c1;
save transaction savePoint5;
insert into t1 values(16);
save transaction savePoint6;
insert into t1 values(17);
save transaction savePoint7;
insert into t1 values(18);
select * from t1 order by c1;
rollback transaction savePoint6;
select * from t1 order by c1;
commit transaction;
select * from t1 order by c1;

begin transaction;
insert into t1 values(19);
save transaction savePoint8;
insert into t1 values(20);
save transaction savePoint9;
insert into t1 values(21);
save transaction savePoint10;
insert into t1 values(22);
select * from t1 order by c1;
rollback transaction savePoint8;
select * from t1 order by c1;
commit transaction;
select * from t1 order by c1;

begin transaction transaction1;
save transaction savePoint1;
insert into t1 values(23);
select * from t1 order by c1;
-- expect error
rollback transaction savePoint2;
select * from t1 order by c1;
rollback transaction;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(24);
begin transaction transaction2;
insert into t1 values(25);
select * from t1 order by c1;
commit transaction transaction2;
select * from t1 order by c1;
commit transaction transaction1;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(26);
begin transaction transaction2;
insert into t1 values(27);
select * from t1 order by c1;
rollback transaction;
select * from t1 order by c1;
rollback transaction;
select * from t1 order by c1;

begin transaction transaction1;
insert into t1 values(28);
begin transaction transaction2;
insert into t1 values(29);
select * from t1 order by c1;
commit transaction transaction1;
select * from t1 order by c1;
commit transaction transaction2;
select * from t1 order by c1;

truncate table t1;
begin transaction;
insert into t1 values(1);
commit work;
select * from t1;

begin transaction;
insert into t1 values(2);
rollback work;
select * from t1;

begin transaction;
insert into t1 values(3);
commit;
select * from t1;

begin transaction;
insert into t1 values(4);
rollback;
select * from t1;

truncate table t1;
create or replace procedure p1 () as 
BEGIN
    insert into t1 values(1);
    commit transaction;
    insert into t1 values(2);
    rollback transaction;
    save tran savepoint1;
    insert into t1 values(3);
    rollback tran savepoint1;
    insert into t1 values(4);
    commit transaction transactionName1;
END;
/

select p1();
select * from t1;

truncate table t1;
create or replace function f1 () returns int language pltsql as 
$$
BEGIN
    insert into t1 values(5);
    commit tran transactionName2;
    insert into t1 values(6);
    rollback tran;
    save tran savepoint1;
    insert into t1 values(7);
    rollback tran savepoint1;
    insert into t1 values(8);
    commit tran;
    return 1;
end;
$$;
select f1();
select * from t1;

truncate table t1;
do $$
BEGIN
    insert into t1 values(1);
    commit transaction;
    insert into t1 values(2);
    rollback transaction;
    save tran savepoint1;
    insert into t1 values(3);
    rollback tran savepoint1;
    insert into t1 values(4);
    commit transaction;
END
$$;


drop table if exists ValueTable;
CREATE TABLE ValueTable (id INT);

create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(1);
rollback;
COMMIT;
end;
/

create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(2);
SAVE TRAN save_point1;
COMMIT;
end;
/

create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(3);
ROLLBACK TRAN;
COMMIT TRAN;
end;
/

create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(4);
ROLLBACK TRANSACTION;
COMMIT TRANSACTION;
end;
/

drop procedure procedure_15();

BEGIN
    INSERT INTO ValueTable VALUES(1);
    SAVE TRAN save_point1;
    INSERT INTO ValueTable VALUES(2);
    ROLLBACK TRAN save_point1;
    INSERT INTO ValueTable VALUES(3);
    COMMIT TRANSACTION;
END;
/

select * from ValueTable;

drop table ValueTable;
drop function p1();
drop function f1();
drop table t1;
drop schema transaction_test cascade;
