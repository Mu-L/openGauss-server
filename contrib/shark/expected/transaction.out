create schema transaction_test;
set search_path = 'transaction_test';
create table t1(c1 int);
begin tran;
insert into t1 values(1);
select * from t1 order by c1;
 c1 
----
  1
(1 row)

commit tran;
select * from t1 order by c1;
 c1 
----
  1
(1 row)

begin tran transaction1;
insert into t1 values(2);
select * from t1 order by c1;
 c1 
----
  1
  2
(2 rows)

commit tran transaction1;
select * from t1 order by c1;
 c1 
----
  1
  2
(2 rows)

begin tran transaction1;
insert into t1 values(3);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
(3 rows)

commit tran transaction2;
begin tran transaction1;
insert into t1 values(4);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
(4 rows)

commit tran;
begin tran;
insert into t1 values(5);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

commit tran transaction2;
begin tran;
insert into t1 values(6);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  6
(6 rows)

rollback tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

begin tran transaction1;
insert into t1 values(7);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  7
(6 rows)

rollback tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

begin tran transaction1;
insert into t1 values(8);
save tran savepoint1;
insert into t1 values(9);
rollback tran savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

commit tran;
begin tran transaction1;
insert into t1 values(10);
save tran savepoint1;
insert into t1 values(11);
rollback tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

begin tran;
save tran savePoint1;
insert into t1 values(12);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 12
(7 rows)

rollback tran savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

save tran savePoint2;
insert into t1 values(13);
save tran savePoint3;
insert into t1 values(14);
save tran savePoint4;
insert into t1 values(15);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
 14
 15
(9 rows)

rollback tran savePoint4;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
 14
(8 rows)

rollback tran savePoint3;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
(7 rows)

rollback tran savePoint2;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

commit tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

begin tran;
save tran savePoint1;
drop table t1;
\d t1;
rollback tran savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

save tran savePoint5;
insert into t1 values(16);
save tran savePoint6;
insert into t1 values(17);
save tran savePoint7;
insert into t1 values(18);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 17
 18
(9 rows)

rollback tran savePoint6;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
(7 rows)

commit tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
(7 rows)

begin tran;
insert into t1 values(19);
save tran savePoint8;
insert into t1 values(20);
save tran savePoint9;
insert into t1 values(21);
save tran savePoint10;
insert into t1 values(22);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 20
 21
 22
(11 rows)

rollback tran savePoint8;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

commit tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

begin tran transaction1;
save tran savePoint1;
insert into t1 values(23);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 23
(9 rows)

-- expect error
rollback tran savePoint2;
ERROR:  no such savepoint
select * from t1 order by c1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block, firstChar[Q]
rollback tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

begin tran transaction1;
insert into t1 values(24);
begin tran transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(25);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

commit tran transaction2;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

commit tran transaction1;
WARNING:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

begin tran transaction1;
insert into t1 values(26);
begin tran transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(27);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 26
 27
(12 rows)

rollback tran;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

rollback tran;
NOTICE:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

begin tran transaction1;
insert into t1 values(28);
begin tran transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(29);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

commit tran transaction1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

commit tran transaction2;
WARNING:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

truncate table t1;
begin tran;
insert into t1 values(1);
commit work;
select * from t1;
 c1 
----
  1
(1 row)

begin tran;
insert into t1 values(2);
rollback work;
select * from t1;
 c1 
----
  1
(1 row)

begin tran;
insert into t1 values(3);
commit;
select * from t1;
 c1 
----
  1
  3
(2 rows)

begin tran;
insert into t1 values(4);
rollback;
select * from t1;
 c1 
----
  1
  3
(2 rows)

truncate table t1;
begin transaction;
insert into t1 values(1);
select * from t1 order by c1;
 c1 
----
  1
(1 row)

commit transaction;
select * from t1 order by c1;
 c1 
----
  1
(1 row)

begin transaction transaction1;
insert into t1 values(2);
select * from t1 order by c1;
 c1 
----
  1
  2
(2 rows)

commit transaction transaction1;
select * from t1 order by c1;
 c1 
----
  1
  2
(2 rows)

begin transaction transaction1;
insert into t1 values(3);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
(3 rows)

commit transaction transaction2;
begin transaction transaction1;
insert into t1 values(4);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
(4 rows)

commit transaction;
begin transaction;
insert into t1 values(5);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

commit transaction transaction2;
begin transaction;
insert into t1 values(6);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  6
(6 rows)

rollback transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

begin transaction transaction1;
insert into t1 values(7);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  7
(6 rows)

rollback transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
(5 rows)

begin transaction transaction1;
insert into t1 values(8);
save transaction savepoint1;
insert into t1 values(9);
rollback transaction savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

commit transaction;
begin transaction transaction1;
insert into t1 values(10);
save transaction savepoint1;
insert into t1 values(11);
rollback transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

begin transaction;
save transaction savePoint1;
insert into t1 values(12);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 12
(7 rows)

rollback transaction savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

save transaction savePoint2;
insert into t1 values(13);
save transaction savePoint3;
insert into t1 values(14);
save transaction savePoint4;
insert into t1 values(15);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
 14
 15
(9 rows)

rollback transaction savePoint4;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
 14
(8 rows)

rollback transaction savePoint3;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 13
(7 rows)

rollback transaction savePoint2;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

commit transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

begin transaction;
save transaction savePoint1;
drop table t1;
\d t1;
rollback transaction savePoint1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
(6 rows)

save transaction savePoint5;
insert into t1 values(16);
save transaction savePoint6;
insert into t1 values(17);
save transaction savePoint7;
insert into t1 values(18);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 17
 18
(9 rows)

rollback transaction savePoint6;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
(7 rows)

commit transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
(7 rows)

begin transaction;
insert into t1 values(19);
save transaction savePoint8;
insert into t1 values(20);
save transaction savePoint9;
insert into t1 values(21);
save transaction savePoint10;
insert into t1 values(22);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 20
 21
 22
(11 rows)

rollback transaction savePoint8;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

commit transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

begin transaction transaction1;
save transaction savePoint1;
insert into t1 values(23);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 23
(9 rows)

-- expect error
rollback transaction savePoint2;
ERROR:  no such savepoint
select * from t1 order by c1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block, firstChar[Q]
rollback transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
(8 rows)

begin transaction transaction1;
insert into t1 values(24);
begin transaction transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(25);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

commit transaction transaction2;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

commit transaction transaction1;
WARNING:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

begin transaction transaction1;
insert into t1 values(26);
begin transaction transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(27);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 26
 27
(12 rows)

rollback transaction;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

rollback transaction;
NOTICE:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
(10 rows)

begin transaction transaction1;
insert into t1 values(28);
begin transaction transaction2;
WARNING:  there is already a transaction in progress
insert into t1 values(29);
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

commit transaction transaction1;
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

commit transaction transaction2;
WARNING:  there is no transaction in progress
select * from t1 order by c1;
 c1 
----
  1
  2
  3
  4
  5
  8
 16
 19
 24
 25
 28
 29
(12 rows)

truncate table t1;
begin transaction;
insert into t1 values(1);
commit work;
select * from t1;
 c1 
----
  1
(1 row)

begin transaction;
insert into t1 values(2);
rollback work;
select * from t1;
 c1 
----
  1
(1 row)

begin transaction;
insert into t1 values(3);
commit;
select * from t1;
 c1 
----
  1
  3
(2 rows)

begin transaction;
insert into t1 values(4);
rollback;
select * from t1;
 c1 
----
  1
  3
(2 rows)

truncate table t1;
create or replace procedure p1 () as 
BEGIN
    insert into t1 values(1);
    commit transaction;
    insert into t1 values(2);
    rollback transaction;
    save tran savepoint1;
    insert into t1 values(3);
    rollback tran savepoint1;
    insert into t1 values(4);
    commit transaction transactionName1;
END;
/
select p1();
 p1 
----
 
(1 row)

select * from t1;
 c1 
----
  1
  4
(2 rows)

truncate table t1;
create or replace function f1 () returns int language pltsql as 
$$
BEGIN
    insert into t1 values(5);
    commit tran transactionName2;
    insert into t1 values(6);
    rollback tran;
    save tran savepoint1;
    insert into t1 values(7);
    rollback tran savepoint1;
    insert into t1 values(8);
    commit tran;
    return 1;
end;
$$;
select f1();
 f1 
----
  1
(1 row)

select * from t1;
 c1 
----
  5
  8
(2 rows)

truncate table t1;
do $$
BEGIN
    insert into t1 values(1);
    commit transaction;
    insert into t1 values(2);
    rollback transaction;
    save tran savepoint1;
    insert into t1 values(3);
    rollback tran savepoint1;
    insert into t1 values(4);
    commit transaction;
END
$$;
drop table if exists ValueTable;
NOTICE:  table "valuetable" does not exist, skipping
CREATE TABLE ValueTable (id INT);
create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(1);
rollback;
COMMIT;
end;
/
create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(2);
SAVE TRAN save_point1;
COMMIT;
end;
/
create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(3);
ROLLBACK TRAN;
COMMIT TRAN;
end;
/
create or replace procedure procedure_15()
is
begin
INSERT INTO ValueTable VALUES(4);
ROLLBACK TRANSACTION;
COMMIT TRANSACTION;
end;
/
drop procedure procedure_15();
BEGIN
    INSERT INTO ValueTable VALUES(1);
    SAVE TRAN save_point1;
    INSERT INTO ValueTable VALUES(2);
    ROLLBACK TRAN save_point1;
    INSERT INTO ValueTable VALUES(3);
    COMMIT TRANSACTION;
END;
/
select * from ValueTable;
 id 
----
  1
  3
(2 rows)

select pg_typeof(@@FETCH_STATUS);
 pg_typeof 
-----------
 integer
(1 row)

DO $$
DECLARE
 fetch_status int;
BEGIN
 fetch_status := @@FETCH_STATUS;
 RAISE NOTICE 'fetch status: %', fetch_status;
END;
$$ LANGUAGE plpgsql;
NOTICE:  fetch status: 0
DO $$
DECLARE
 fetch_status text;
BEGIN
 fetch_status := @@FETCH_STATUS;
 RAISE NOTICE 'fetch status: %', fetch_status;
END;
$$ LANGUAGE plpgsql;
NOTICE:  fetch status: 0
DO $$
DECLARE
 fetch_status bool;
BEGIN
 fetch_status := @@FETCH_STATUS;
 RAISE NOTICE 'fetch status: %', fetch_status;
END;
$$ LANGUAGE plpgsql;
NOTICE:  fetch status: f
DO $$
DECLARE
 fetch_status json;
BEGIN
 fetch_status := @@FETCH_STATUS;
 RAISE NOTICE 'fetch status: %', fetch_status;
END;
$$ LANGUAGE plpgsql;
NOTICE:  fetch status: 0
DO $$
DECLARE
 fetch_status date;
BEGIN
 fetch_status := @@FETCH_STATUS;
 RAISE NOTICE 'fetch status: %', fetch_status;
END;
$$ LANGUAGE plpgsql;
ERROR:  date/time field value out of range: "0"
HINT:  Perhaps you need a different "datestyle" setting.
CONTEXT:  PL/pgSQL function inline_code_block line 5 at assignment
drop table ValueTable;
drop function p1();
drop function f1();
drop table t1;
drop schema transaction_test cascade;
